internal/storage/dropbox/dropbox.go

```diff
@@ -0,0 +1,267 @@
+package dropbox
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox"
+	"github.com/dropbox/dropbox-sdk-go-unofficial/v6/dropbox/files"
+	"github.com/offen/docker-volume-backup/internal/storage"
+	"golang.org/x/oauth2"
+)
+
+type dropboxStorage struct {
+	*storage.StorageBackend
+	client           files.Client
+	concurrencyLevel int
+}
+
+// Config allows to configure a Dropbox storage backend.
+type Config struct {
+	Endpoint         string
+	RefreshToken     string
+	AppKey           string
+	AppSecret        string
+	RemotePath       string
+	ConcurrencyLevel int
+}
+
+// NewStorageBackend creates and initializes a new Dropbox storage backend.
+func NewStorageBackend(opts Config, logFunc storage.Log) (storage.Backend, error) {
+	tokenUrl, _ := url.JoinPath(opts.Endpoint, "oauth2/token")
+
+	conf := &oauth2.Config{
+		ClientID:     opts.AppKey,
+		ClientSecret: opts.AppSecret,
+		Endpoint: oauth2.Endpoint{
+			TokenURL: tokenUrl,
+		},
+	}
+
+	isCITest := opts.Endpoint != "https://api.dropbox.com/"
+
+	logFunc(storage.LogLevelInfo, "Dropbox", "Fetching fresh access token for Dropbox storage backend.")
+	token := &oauth2.Token{RefreshToken: opts.RefreshToken}
+	if !isCITest {
+		tkSource := conf.TokenSource(context.TODO(), &oauth2.Token{RefreshToken: opts.RefreshToken})
+		var err error
+		token, err = tkSource.Token()
+		if err != nil {
+			return nil, fmt.Errorf("(*dropboxStorage).NewStorageBackend: Error refreshing token: %w", err)
+		}
+	}
+
+	dbxConfig := dropbox.Config{}
+
+	if isCITest {
+		dbxConfig.Token = opts.RefreshToken
+		dbxConfig.URLGenerator = func(hostType string, namespace string, route string) string {
+			return fmt.Sprintf("%s/%d/%s/%s", opts.Endpoint, 2, namespace, route)
+		}
+	} else {
+		dbxConfig.Token = token.AccessToken
+	}
+
+	client := files.New(dbxConfig)
+
+	if opts.ConcurrencyLevel < 1 {
```

Comment:
Good idea to check this. I would even go as far as returning an error here, so people are actually forced to provide a valid configuration instead of getting numbed by warnings.

Reply:
Alternatively, we could provide a custom `NaturalNumber` type catch such cases at config level already:

```go
type NaturalNumber int

func (n *NaturalNumber) Decode(v string) error {
	asInt, err := strconv.Atoi(v)
	if err != nil {
		return fmt.Errorf("config: error converting %s to int", v)
	}
	if asInt <= 0 {
		return fmt.Errorf("config: expected a natural number, got %d", asInt)
	}
	*n = NaturalNumber(asInt)
	return nil
}

func (n *NaturalNumber) Int() int {
	return int(*n)
}

type Config struct {
	DropboxConcurrencyLevel NaturalNumber `split_words:"true" default:"6"`
}
```

Reply:
Good idea, the configs could benefit in general from that.